rust/kernel/chrdev.rs

```diff
@@ -20,10 +19,57 @@ use crate::error::{Error, KernelResult};
 use crate::file_operations;
 use crate::types::CStr;
 
+/// Character device.
+///
+/// # Invariants
+///
+/// The pointer [`Cdev::cdev`] is valid and non-null.
+struct Cdev(*mut bindings::cdev);
+
+impl Cdev {
+    fn alloc(
+        fops: &'static bindings::file_operations,
+        module: &'static crate::ThisModule,
+    ) -> KernelResult<Self> {
+        // SAFETY: `bindings::cdev_alloc` returns a valid pointer or null.
+        let cdev = unsafe { bindings::cdev_alloc() };
+        if cdev.is_null() {
+            return Err(Error::ENOMEM);
+        }
+        // SAFETY: [`Cdev::cdev`] is valid and non-null.
```

Comment:
Here we don't have a `Cdev` object yet, so the type invariants do not guarantee the pointer is valid.

Instead, consider something like:

```suggestion
        // SAFETY: `cdev` should be valid since `cdev_alloc()` returned a non-null pointer.
```